relayer/chains/evm/listener.go

```diff
@@ -81,120 +70,82 @@ func (p *Provider) Listener(ctx context.Context, lastSavedHeight uint64, blockIn
 		case <-subscribeStart.C:
 			subscribeStart.Stop()
 			go p.Subscribe(ctx, blockInfoChan, resetFunc)
-		case <-pollHeightTicker.C:
-			pollHeightTicker.Stop()
-			next = p.GetLastSavedBlockHeight()
-			latest = p.latestHeight()
-
-		case bn := <-bnch:
-			// process all notifications
-			for ; bn != nil; next++ {
-				if lbn != nil {
-					p.log.Debug("block-notification received", zap.Uint64("height", lbn.Height.Uint64()),
-						zap.Int64("gas-used", int64(lbn.Header.GasUsed)))
-
-					messages, err := p.FindMessages(ctx, lbn)
-					if err != nil {
-						return errors.Wrapf(err, "receiveLoop: callback: %v", err)
-					}
-					blockInfoChan <- &relayertypes.BlockInfo{
-						Height:   lbn.Height.Uint64(),
-						Messages: messages,
-					}
-				}
 
-				if lbn, bn = bn, nil; len(bnch) > 0 {
-					bn = <-bnch
-				}
-			}
-			// remove unprocessed notifications
-			for len(bnch) > 0 {
-				<-bnch
-			}
-
-		default:
-			if next >= latest {
-				continue
+			if isSubError {
+				startHeight = p.GetLastSavedBlockHeight()
 			}
 
-			qch := make(chan *bnq, cap(bnch))
-			for i := next; i < latest && len(qch) < cap(qch); i++ {
-				qch <- &bnq{i, nil, nil, 3} // fill bch with requests
+			var blockReqs []*blockReq
+			for i := startHeight; i < latestHeight; i += p.cfg.BlockBatchSize {
+				end := i + p.cfg.BlockBatchSize
```

Comment:
Here, the for loop will generate one duplicate block query always.
Suppose startHeight = 1, batchSize=3 and latestHeight=10
in this case  as per this implementation:
[{1, 4}, {4,7},{7,10}] is generated. But I think we need [{1, 3}, {4,6},{7,9},{10,10}]

Reply:
yeah, nice cactch.